src/pybind/mgr/diskprediction/module.py

```diff
@@ -0,0 +1,346 @@
+                            'disk_wwn': dev.get('diskWWN'),
+                            'attachment': p_data.get('disk_name', '')
+                        }
+                        physical_data[dev.get('diskName')] = d_data
+                    osd_data['osd.%s' % cmd['osd_id']] = dict(
+                        prediction=physical_data)
+                    break
+            if not osd_data:
+                msg = 'not found osd %s predicted data' % cmd['osd_id']
+            else:
+                msg = json.dumps(osd_data, indent=4)
+        except Exception as e:
+            if str(e).find('No such file') >= 0:
+                msg = 'unable to get osd %s predicted data'
+            else:
+                msg = 'unable to get osd %s predicted data, %s' % (cmd['osd_id'], str(e))
+            self.log.error(msg)
+        return 0, msg, ''
+
+    def _config_set(self, inbuf, cmd):
+        self.set_config('diskprediction_server', cmd['server'])
+        self.set_config('diskprediction_user', cmd['user'])
+        self.set_config('diskprediction_password', cmd['password'])
+        if cmd.get('port'):
+            self.set_config('diskprediction_port', cmd['port'])
+        self.show_module_config()
+        return 0, json.dumps(self.config, indent=4), ''
+
+    def _run_prediction_forced(self, inbuf, cmd):
+        msg = ''
+        for _task in self._tasks:
+            if isinstance(_task, PredictionTask):
+                msg = 'run prediction agent successfully'
+                _task.event.set()
+        return 0, msg, ''
+
+    def _run_metrics_forced(self, inbuf, cmd):
+        msg = ''
+        for _task in self._tasks:
+            if isinstance(_task, MetricsTask):
+                msg = 'run metrics agent successfully'
+                _task.event.set()
+        return 0, msg, ''
+
+    def _run_smart_forced(self, inbuf, cmd):
+        msg = ' '
+        for _task in self._tasks:
+            if isinstance(_task, SmartTask):
+                msg = 'run smart agent successfully'
+                _task.event.set()
+        return 0, msg, ''
+
+    def _status(self, inbuf, cmd):
+        msg = 'diskprediction plugin status: %s' % self.status
+        return 0, msg, ''
+
+    def handle_command(self, inbuf, cmd):
+        for o_cmd in self.COMMANDS:
+            if cmd['prefix'] == o_cmd['cmd'][:len(cmd['prefix'])]:
+                fun = getattr(
+                    self, '_%s' % o_cmd['cmd'].split(' ')[1].replace('-', '_'))
+                if fun:
+                    return fun(inbuf, cmd)
+        return 0, 'cmd not found', ''
+
+    def show_module_config(self):
+        self.fsid = self.get('mon_map')['fsid']
+        self.log.debug('Found Ceph fsid %s', self.fsid)
+
+        for key, default in self.config_keys.items():
+            self.set_config_option(key, self.get_config(key, default))
+
+    def serve(self):
+        self.log.info('Starting diskprediction module')
+        self.run = True
```

Comment:
It's racy to set self.run here, as it's possible that shutdown() was already called.  Looks like this bug may also exist in other modules.

You don't need a separate self.run variable: just rename self.event to self.shutdown_event, and use self.shutdown_event.is_set as the indication to drop out of this loop.